src/psbt.cpp

```diff
@@ -466,36 +466,41 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
     return sig_complete ? PSBTError::OK : PSBTError::INCOMPLETE;
 }
 
-void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx, const int& sighash_type)
+void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx)
 {
-    // Only drop non_witness_utxos if sighash_type != SIGHASH_ANYONECANPAY
-    if ((sighash_type & 0x80) != SIGHASH_ANYONECANPAY) {
-        // Figure out if any non_witness_utxos should be dropped
-        std::vector<unsigned int> to_drop;
-        for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
-            const auto& input = psbtx.inputs.at(i);
-            int wit_ver;
-            std::vector<unsigned char> wit_prog;
-            if (input.witness_utxo.IsNull() || !input.witness_utxo.scriptPubKey.IsWitnessProgram(wit_ver, wit_prog)) {
-                // There's a non-segwit input or Segwit v0, so we cannot drop any witness_utxos
-                to_drop.clear();
-                break;
-            }
-            if (wit_ver == 0) {
-                // Segwit v0, so we cannot drop any non_witness_utxos
-                to_drop.clear();
-                break;
-            }
-            if (input.non_witness_utxo) {
-                to_drop.push_back(i);
-            }
+    // Figure out if any non_witness_utxos should be dropped
+    std::vector<unsigned int> to_drop;
+    for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
+        const auto& input = psbtx.inputs.at(i);
+        int wit_ver;
+        std::vector<unsigned char> wit_prog;
+        if (input.witness_utxo.IsNull() || !input.witness_utxo.scriptPubKey.IsWitnessProgram(wit_ver, wit_prog)) {
+            // There's a non-segwit input or Segwit v0, so we cannot drop any witness_utxos
+            to_drop.clear();
+            break;
+        }
+        if (wit_ver == 0) {
+            // Segwit v0, so we cannot drop any non_witness_utxos
+            to_drop.clear();
+            break;
+        }
+        // non_witness_utxos can only be dropped if the sighash type does not include SIGHASH_ANYONECANPAY
```

Comment:
In 6ec298a335585ef723d113b94959377d3ff65759 "psbt: use sighash type field to determine whether to remove non-witness utxos":

"can only ... if not" is hard to parse, and it seems the code doing the opposite. It also seems to do the opposite of the original behavior.

```
// Only drop non_witness_utxos if sighash_type != SIGHASH_ANYONECANPAY
```

Can you clarify the reason why `SIGHASH_ANYONECANPAY` requires different treatment in the first place?

Reply:
Fixed and reworded the comment.

`SIGHASH_ANYONECANPAY` does not commit to any input information, so none of the protections that allow us to drop the `non_witness_utxo` apply.

Reply:
Ok, in 795e911ded2563705cec69e43ff622ace22d58d5 "psbt: use sighash type field to determine whether to remove non-witness utxos" it's now more clear that this just moves a check inside the `for` loop.

I still can't wrap my head around the logic of how and why we drop `non_witness_utxos`.